lib/Sema/CSApply.cpp

```diff
@@ -7005,12 +7005,30 @@ Expr *ExprRewriter::finishApply(ApplyExpr *apply, Type openedType,
     return apply;
   }
 
-  // If the metatype value isn't a type expression, the user should reference
-  // '.init' explicitly, for clarity.
   if (!cs.isTypeReference(fn)) {
-    cs.TC.diagnose(apply->getArg()->getStartLoc(),
-                   diag::missing_init_on_metatype_initialization)
-      .fixItInsert(apply->getArg()->getStartLoc(), ".init");
+    auto fnType = fn->getType();
+    auto &TC = cs.TC;
+
+    // If this is an attempt to initialize existential type.
+    if (auto metaType = fnType->getAs<MetatypeType>()) {
+      auto instanceType = metaTy->getInstanceType();
+      if (instanceType->isExistentialType()) {
+        auto diagnostic = fn->isStaticallyDerivedMetatype()
+                              ? diag::construct_protocol_by_name
+                              : diag::construct_protocol_value;
+
+        TC.diagnose(fn->getLoc(), diagnostic, metaTy)
+            .highlight(fn->getSourceRange());
+        return nullptr;
```

Comment:
Ok, I've checked getArgCount and it already has special handling for the situation you mentioned, and I've figured why matchCallArguments fails - it's because the locator `init` in this case is - `apply function -> constructor member` but coerceCallArguments only works with `apply function` locators, and because of that fails to determine callee declaration at `CSApply.cpp` L5315 so when we go ahead and match arguments matcher doesn't know that there is a parameter with default value and fails with missing argument...